modules/katamari/src/main/ts/ephox/katamari/api/Result.ts

```diff
@@ -1,156 +1,288 @@
+ */
+interface DebugResult<T, E> extends Result<T, E> {
+  /**
+   * True for `Value`, False for `Error`.
+   */
+  readonly tag: boolean;
 
-  const fold = <U>(onError: (err: E) => U, _: (value: T) => U) => {
-    return onError(message);
+  /**
+   * The value or error contained within this Result object.
+   */
+  readonly inner: T | E;
+}
+
+/**
+ * Creates a new `Result<T, E>` that **does** contain a value.
+ */
+const value = <T, E = never>(value: T): Result<T, E> => {
+
+  const applyHelper = <U>(fn: (value: T) => U): U => fn(value);
+  const constHelper = Fun.constant(value);
+  const outputHelper = <E2 = E>() => output as unknown as Result<T, E2>;
+
+  const output: DebugResult<T, E> = {
+    // Debug info
+    tag: true,
+    inner: value,
+    // Actual Result methods
+    fold: (_onError, onValue) => onValue(value),
+    isValue: Fun.always,
+    isError: Fun.never,
+    map: (mapper) => Result.value(mapper(value)),
+    mapError: outputHelper,
+    bind: applyHelper,
+    exists: applyHelper,
+    forall: applyHelper,
+    getOr: constHelper,
+    or: outputHelper,
+    getOrThunk: constHelper,
+    orThunk: outputHelper,
+    getOrDie: constHelper,
+    each: (fn) => {
+      // Can't write the function inline because we don't want to return something by mistake
+      fn(value);
+    },
+    toOptional: () => Optional.some(value),
   };
 
-  return {
+  return output;
+};
+
+/**
+ * Creates a new `Result<T, E>` that **does not** contain a value, and therefore
+ * contains an error.
+ */
+const error = <T = never, E = any>(error: E): Result<T, E> => {
+
+  const output: DebugResult<T, E> = {
+    // Debug info
+    tag: false,
+    inner: error,
+    // Actual Result methods
+    fold: (onError, _onValue) => onError(error),
     isValue: Fun.never,
     isError: Fun.always,
-    getOr: Fun.identity,
-    getOrThunk,
-    getOrDie,
-    or,
-    orThunk,
-    fold,
-    map,
-    mapError,
-    each: Fun.noop,
-    bind,
+    map: <U>(_mapper) => output as unknown as Result<U, E>,
+    mapError: (mapper) => Result.error(mapper(error)),
+    bind: <U>(_binder) => output as unknown as Result<U, E>,
     exists: Fun.never,
     forall: Fun.always,
-    toOptional: Optional.none
+    getOr: Fun.identity,
+    or: Fun.identity,
+    getOrThunk: Fun.apply,
+    orThunk: Fun.apply,
+    getOrDie: Fun.die(String(error)),
+    each: Fun.noop,
+    toOptional: Optional.none,
   };
+
+  return output;
 };
 
-const fromOption = <T, E>(opt: Optional<T>, err: E): Result<T, E> => opt.fold(
-  () => error(err),
-  value
-);
+/**
+ * Creates a new `Result<T, E>` from an `Optional<T>` and an `E`. If the
+ * `Optional` contains a value, so will the outputted `Result`. If it does not,
+ * the outputted `Result` will contain an error (and that error will be the
+ * error passed in).
+ */
+const fromOption = <T, E>(optional: Optional<T>, error: E): Result<T, E> =>
+  optional.fold(() => Result.error(error), Result.value);
```

Comment:
This references variables before they are defined, which we generally try to avoid... that's why the old version didn't do this, so are we sure we want to make this change?